jupyterhub/proxy.py

```diff
@@ -39,10 +39,16 @@
 from traitlets import Unicode
 from traitlets.config import LoggingConfigurable
 
+try:
+    from tornado.curl_httpclient import CurlError
+except ImportError:
+    CurlError = NotImplementedError
+
```

Comment:
This will only work if pycurl is installed, and is only a relevant error to catch if it is installed. By letting CurlError become a reference to NotImplementedError I can just do `except (TimeoutError, CurlError) as error:` later without dynamically creating a tuple etc.

Reply:
Why do you want to attempt to catch a `CurlError` instead of `HTTPClientError` which is used by all HTTP clients?

Reply:
I want to catch specifically Broken Pipe errors which are low level network error (CurlError), while I don't want to catch for example 401 Forbidden which is just a plain failure to authenticate and doesn't make sense to try again.

Does HTTPClientError include 401 errors for example? I wonder exactly what errors we should aim to retry for, it is not obvious in my mind. Some 4XX errors like 429 about "too many requests atm, retry again" would make sense to retry, but I'm not sure if CHP emits those ever.

I'm open to suggestions on what to catch, but I think its sensible to aim for the errors that represents network connection related errors to begin with, how do we single them out?

Reply:
I think the recommendation is to use the client independent classes when catching exceptions. Somewhere Min even commented on this but I can't find the issue/PR :-/

What I'd do is catch `HTTPClientError` and then check the status code. 599 is a special status code used by Tornado to signal "something went wrong and it isn't a normal HTTP status code". For example network request timeouts, broken pipes, etc. That way we can decide for which status codes or range of status codes we do want to retry.

Reply:
Excellent push in the right direction I think @betatim! I'll do this!

599 is after all...
![image](https://user-images.githubusercontent.com/3837114/97203410-109a5000-17b5-11eb-9aa6-87ed3e555eec.png)

Reply:
```python
class CurlError(HTTPError):
    def __init__(self, errno: int, message: str) -> None:
        HTTPError.__init__(self, 599, message)
        self.errno = errno
```
- [Source code for CurlError in tornado here](https://github.com/tornadoweb/tornado/blob/bee08e25b1b990edd7ebfb7cd463ec18a98a3fde/tornado/curl_httpclient.py#L575-L578)

```python
class HTTPError(Exception):
    """An exception that will turn into an HTTP error response.
    ...
    """
```
- [Source code for HTTPError in tornado here](https://github.com/tornadoweb/tornado/blob/bee08e25b1b990edd7ebfb7cd463ec18a98a3fde/tornado/web.py#L2333-L2338)

---

It seems like HTTPClientError is the same as HTTPError.

![image](https://user-images.githubusercontent.com/3837114/97218212-adb2b400-17c8-11eb-9ecd-6d6ac9e14f99.png)

---

Conclusion, I should catch HTTPError / HTTPClientError, but raise it again if it doesn't seem appropriate to handle, and in this case I'm currently only confident about 599 errors.

Reply:
I found this example from the existing code base.

```python
        except HTTPError as e:
            if e.code == 404:
                # Warn about 404s because something might be wrong
                # but don't raise because the route is gone,
                # which is the goal.
                self.log.warning("Route %s already deleted", routespec)
            else:
                raise
```

Thanks @betatim!

Reply:
Resolved in b91c5a4

Reply:
I also made 429,500,503,504 retry as well, which is in line with recommendations for the k8s api-server, so it probably can make some general sense I think.

See this related discussion: https://github.com/jupyterhub/kubespawner/issues/436